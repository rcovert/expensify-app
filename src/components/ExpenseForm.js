import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// old way of date manipulation const date = new Date();
// const now = moment();  // returns now
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        // set values to existing expense values for edit
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        };
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description: description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        // updater function (() => ({object}))
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        // validate amount
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        };
    };
    onDateChange = (createdAt) => {
        // check if there is a value - if not do nothing
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        // avoid full page refresh behavior 
        e.preventDefault();
        // perform validation on desc and amt
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please add a description and amount to the form.' }));
        } else {
            this.setState(() => ({ error: '' }));
            // now pass the valid data back to parent component for action dispatch
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    // react class components must have render method
    // note use of form as root element for page
    // does not have to be a div
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type='text'
                    className="text_input"
                    placeholder='Description'
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type='text'
                    className="text_input"
                    placeholder='Amount'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    placeholder='Add a note for your expense (optional)'
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}>
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>

            </form>
        )
    }
};
