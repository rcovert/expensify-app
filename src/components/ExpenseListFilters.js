import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

// note: connected objects have direct access to, dispatch function through props
// e is the event on the input line and has access to what user types
export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) => {
        // updater function wraps an object in parent
        this.setState(() => ({ calendarFocused }));
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onFilterChange = (e) => {
        e.target.value === 'date' ?
            this.props.sortByDate() :
            this.props.sortByAmount()
    }
    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={this.onTextChange} />
                <select value={this.props.filters.sortBy}
                    onChange={this.onFilterChange}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select >
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    showClearDates={true}
                    isOutsideRange={() => false}
                />

            </div >
        );
    };
};

const mapStateToProps = (state) => ({
    // returning an object - so need {} notation
        filters: state.filters
});

const mapDispatchToProps = (dipatch) => ({
    // composed of an object containing arrow functions assigned to actions
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate),
    sortByAmount: () => dispatch(sortByAmount),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))

});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);