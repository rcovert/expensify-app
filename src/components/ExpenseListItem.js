import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

export const ExpenseListItem = ({ id, description, amount, createdAt }) => {

    return (
        // note: shorthand of id: id -> is just id
        <div>
            <Link
                to={`/edit/${id}`} >
                <h3>{description}</h3>
            </Link>
            <p>{amount} - {createdAt}</p>
        </div>
    );
};

export default ExpenseListItem;
