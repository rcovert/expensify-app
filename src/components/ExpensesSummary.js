import React from 'react';
import { connect, Connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectTotalExpenses from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => (
    <div>
        {
            expenseCount === 0 ? (
                <p>No Expense Items</p>
            ) : (
                <p>There are {expenseCount} item(s), totaling:  
                  {numeral(expensesTotal/100).format(' $0,0.00')}</p>
            )
        }
    </div>
)

const mapStateToProps = (state) => {
    // pass the expenses array as a prop/expenses to the expense list
    // setup visibleExpenses as viewed array
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    // create object with new props to display in jsx above
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectTotalExpenses(visibleExpenses)
    };
};

// connect returns a function that then is called with ExpenseList to connect the
// component to the store
export default connect(mapStateToProps)(ExpensesSummary);