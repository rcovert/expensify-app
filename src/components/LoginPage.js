import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({startLogin}) => {
    return (<div>
        <p>Please Login to Expensify</p>
        <button onClick={startLogin}>Login</button>
    </div>)

};

const mapDispathToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispathToProps)(LoginPage);