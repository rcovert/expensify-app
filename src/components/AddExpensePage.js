import React from "react";
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

// note props are passed in from connected object 

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // now dispath redux action console.log(expense);
        // now route back to the dashboard with history.push method from router/route
        // props.dispatch(addExpense(expense));
        // use this.props for class based components - not required on stateless functional components
        this.props.addExpense(expense);
        this.props.history.push('/');
    };
    render() {
        // note - this form of the code avoids writing function in-line
        // which prevents recalculation of function on every render
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    };
};

// stateless functional version
// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit={(expense) => {
//                 props.onSubmit(expense);
//                 props.history.push('/');
//             }}
//         />
//     </div>
// );

// mapDispatchToProps is used to simplify code and make it easier for testing
const mapDispatchToProps = (dispatch) => ({
    // implicit return of object -- no need for return statement
    addExpense: (expense) => dispatch(addExpense(expense))
});
// mapStateToProps is first arg; mapDispatchToProps is 2nd arg
export default connect(undefined, mapDispatchToProps)(AddExpensePage);