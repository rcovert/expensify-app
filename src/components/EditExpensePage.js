import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';

// refactor to class based component
// setup mapDispatchToProps

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <p>The id is: {this.props.expense.id}</p>
                <ExpenseForm expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>
                    Remove</button>
            </div>
        )
    }
}

// const EditExpensePage = (props) => {
//     // console.log('the state at edit = ');
//     // console.log(props);
//     return (
//         <div>
//             <h1>Edit Expense</h1>
//             <p>The id is: {props.match.params.id}</p>
//             <ExpenseForm expense={props.expense}
//                 onSubmit={(expense) => {
//                     // now dispath redux action console.log(expense);
//                     // now route back to the dashboard with history.push method from router/route
//                     // console.log(expense.description);
//                     // console.log(props.match.params.id)
//                     props.dispatch(editExpense(props.match.params.id, expense));
//                     props.history.push('/');
//                 }}
//             />
//             <button onClick={() => {
//                 // console.log(props.match.params.id)
//                 // note that we are sending an object to remove expense
//                 props.dispatch(removeExpense({ id: props.match.params.id }))
//                 props.history.push('/');
//             }}>
//                 Remove</button>
//         </div>
//     )
// };

// this is where we need a map state to props function to get the expense with the right id
// which comes in on the match prop
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    // implicit return of object -- no need for return statement
    // edit expense is set to a function
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
});
// need to pass in mapStateToProps to connect to get the props set
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);