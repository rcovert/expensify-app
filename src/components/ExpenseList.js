import React from 'react';
import { connect, Connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// note have to export this component for testing -- in addition to the
// default connected export below
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expense Items</p>
            ) : (
                props.expenses.map((expense) => (
                    <ExpenseListItem key={expense.id} {...expense} />
                ))
            )
        }
        
    </div>
)

const mapStateToProps = (state) => {
    // pass the expenses array as a prop/expenses to the expense list
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

// connect returns a function that then is called with ExpenseList to connect the
// component to the store
export default connect(mapStateToProps)(ExpenseList);
