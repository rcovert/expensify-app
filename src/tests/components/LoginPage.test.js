import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

test('should render Login page correctly', () => {
    const wrapper = shallow(<LoginPage />);
    expect(wrapper).toMatchSnapshot();
});
test('should call startLogin on button press', () => {
    // create a startLogin spy to dummy the function call
    const startLogin = jest.fn();
    const wrapper = shallow(<LoginPage startLogin={startLogin}/>);
    // now simulate click event
    wrapper.find('button').simulate('click');
    expect(startLogin).toHaveBeenCalled();
});