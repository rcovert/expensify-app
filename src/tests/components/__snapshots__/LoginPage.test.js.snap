// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login page correctly 1`] = `
<div>
  <p>
    Please Login to Expensify
  </p>
  <button>
    Login
  </button>
</div>
`;
